apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/EntrySpan.java

```diff
@@ -61,6 +62,14 @@ public EntrySpan tag(String key, String value) {
         return this;
     }
 
+    @Override
+    public AbstractTracingSpan tag(AbstractTag<?> tag, String value) {
+        if (stackDepth == currentMaxDepth || tag.isCanOverwrite() || isInAsyncMode) {
```

Comment:
Maybe 👇 ?

```suggestion
        if (tag.isCanOverwrite() && (stackDepth == currentMaxDepth || isInAsyncMode)) {
```

Reply:
super#tag includes `isCanOverwrite` already. I think we don't need it.

`stackDepth == currentMaxDepth || isInAsyncMode` are the only required conditions.

Reply:
> super#tag includes `isCanOverwrite` already. I think we don't need it.
> 
> `stackDepth == currentMaxDepth || isInAsyncMode` are the only required conditions.

I suppose we still need `isCanOverwrite` check in the `EntrySpan`. `http.status_code` will hit this condition.

Reply:
> > super#tag includes `isCanOverwrite` already. I think we don't need it.
> > `stackDepth == currentMaxDepth || isInAsyncMode` are the only required conditions.
> 
> I suppose we still need `isCanOverwrite` check in the `EntrySpan`. `http.status_code` will hit this condition.

The parent class `AbstractTracingSpan.java#tag` already checks `isCanOverwrite` so we don't need to check again here.

https://github.com/apache/skywalking-java/blob/f04866cf240ace47fa039fd5b0f548f68fb7bee6/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java#L127-L134

Reply:
But in `ExitSpan`

https://github.com/apache/skywalking-java/blob/f04866cf240ace47fa039fd5b0f548f68fb7bee6/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/ExitSpan.java#L67-L72

Is this also redundent?

Reply:
redundant should not represent any bug. Otherwise, it is not redundant at all.
Could you explain your point?

Reply:
Let's leave the bug of what it is.
In this PR, we only need to fix tag set doesn't follow stack depth.

Reply:
> Let's leave the bug of what it is. In this PR, we only need to fix tag set doesn't follow stack depth.

Make sense.

> redundant should not represent any bug. Otherwise, it is not redundant at all. Could you explain your point?

I do not pretty understand the original protocol. But I suppose the critical point is the priority of the `overwritable` property and the `stack depth` of the `Span`. Specifically,

1. For `EntrySpan`, does the "overwritable" tag can be overwritten even if we are not at the top of the stack?
2. For `ExitSpan`, does the "overwritable" tag can be overwritten even if we are not at the bottom of the stack?

Reply:
Think in this way
About entry Span, which means an incoming RPC, we could have Tomcat -> Spring. We need the information on SpringMVC provided, that is why we asked for the deepest. Even it will have more, Tomcat -> Spring -> xxx. Then before xxx happens, Spring will override Tomcat, but eventually, xxx will override Spring.
When we move to the client side(exit span), it is reverted. Then we need the first nested span. (For example), RPC(dubbo) client -> HTTPClient -> netty, we need RPC framework's information.

Reply:
> Think in this way About entry Span, which means an incoming RPC, we could have Tomcat -> Spring. We need the information on SpringMVC provided, that is why we asked for the deepest. Even it will have more, Tomcat -> Spring -> xxx. Then before xxx happens, Spring will override Tomcat, but eventually, xxx will override Spring. When we move to the client side(exit span), it is reverted. Then we need the first nested span. (For example), RPC(dubbo) client -> HTTPClient -> netty, we need RPC framework's information.

Yes. I understand how the stack works. But suppose we have an interceptor between Spring and Tomat,

```
Tomcat(before) -> Spring (before) -> [business code] -> Spring (after) -> [user-defined interceptor] -> Tomcat (after)
                                                             ^
                                                             |
                                                             |
                                                             |
                                                    set http.staus_code
```

It is possible that in the `[user-defined interceptor]`, e.g. could be a [`GenericFilterBean`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/filter/GenericFilterBean.html) in this case, a new response code is set. So I guess that is where `tag.isCanOverwrite()` works.

So I think we should keep `tag.isCanOverwrite()` in both `EntrySpan` and `ExitSpan`.

Reply:
That is right. But you didn't follow what I said. I mean this logic is included when you call `super#tag`. I didn't mean this is wrong. Read `AbstractTracingSpan#tag`

Reply:
> That is right. But you didn't follow what I said. I mean this logic is included when you call `super#tag`. I didn't mean this is wrong. Read `AbstractTracingSpan#tag`

I'm confused

> stackDepth == currentMaxDepth || isInAsyncMode are the only required conditions.

If this condition does not meet, how could the super method been called?

Reply:
An override tag doesn't mean you could override without depth. That is not from the design perspective. 
Override tag means Spring could override Tomcat. It isn't reverted.

Reply:
@lujiajing1126 BTW, why are we talking about this here? I don't think this is relative to this bug at all.

Reply:
> @lujiajing1126 BTW, why are we talking about this here? I don't think this is relative to this bug at all.

I just want to confirm the original design. But for the `ExitSpan` case, we always allow overwrite? Is that a design-oriented decision as well?

Though I suppose there are still limitations, such as the `GenericBeanFilter` case I mentioned above...